subprojects/core/src/main/java/org/gradle/process/internal/DefaultExecAction.java

```diff
@@ -35,117 +36,136 @@
 public class DefaultExecAction implements ExecAction {
 
     private final ClientExecHandleBuilder execHandleBuilder;
+    private final ExecSpec execSpec;
     private boolean ignoreExitValue;
 
-    public DefaultExecAction(ClientExecHandleBuilder execHandleBuilder) {
+    public DefaultExecAction(ExecSpec execSpec, ClientExecHandleBuilder execHandleBuilder) {
+        this.execSpec = execSpec;
         this.execHandleBuilder = execHandleBuilder;
     }
 
     @Override
     public ExecResult execute() {
-        ExecHandle execHandle = execHandleBuilder.build();
+        ExecHandle execHandle = buildExecHandle();
         ExecResult execResult = execHandle.start().waitForFinish();
         if (!isIgnoreExitValue()) {
             execResult.assertNormalExitValue();
         }
         return execResult;
     }
 
+    private ExecHandle buildExecHandle() {
+        execHandleBuilder
+            .commandLine(execSpec.getCommandLine())
+            .setWorkingDir(execSpec.getWorkingDir())
+            .setEnvironment(execSpec.getEnvironment());
+        if (execSpec.getStandardInput() != null) {
+            execHandleBuilder.setStandardInput(execSpec.getStandardInput());
+        }
+        if (execSpec.getStandardOutput() != null) {
+            execHandleBuilder.setStandardOutput(execSpec.getStandardOutput());
+        }
+        if (execSpec.getErrorOutput() != null) {
+            execHandleBuilder.setErrorOutput(execSpec.getErrorOutput());
+        }
+        return execHandleBuilder.build();
+    }
+
     @Override
     public String getExecutable() {
-        return execHandleBuilder.getExecutable();
+        return execSpec.getExecutable();
```

Comment:
💭 We delegate to ExecSpec now, so we have consistent resolution for parameters for all `ExecActions/ExecSpecs` in our code